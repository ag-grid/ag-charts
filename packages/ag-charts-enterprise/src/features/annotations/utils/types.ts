import type { ChannelPropertiesType, LinePropertiesType, TextualPropertiesType } from '../annotationsSuperTypes';
import { CalloutProperties } from '../callout/calloutProperties';
import { CommentProperties } from '../comment/commentProperties';
import { HorizontalLineProperties, VerticalLineProperties } from '../cross-line/crossLineProperties';
import { DisjointChannelProperties } from '../disjoint-channel/disjointChannelProperties';
import { ArrowProperties, LineProperties } from '../line/lineProperties';
import { NoteProperties } from '../note/noteProperties';
import { ParallelChannelProperties } from '../parallel-channel/parallelChannelProperties';
import { TextProperties } from '../text/textProperties';

export function isLineType(datum: unknown): datum is LinePropertiesType {
    return (
        LineProperties.is(datum) ||
        HorizontalLineProperties.is(datum) ||
        VerticalLineProperties.is(datum) ||
        ArrowProperties.is(datum)
    );
}

export function isChannelType(datum: unknown): datum is ChannelPropertiesType {
    return DisjointChannelProperties.is(datum) || ParallelChannelProperties.is(datum);
}

export function isTextType(datum: unknown): datum is TextualPropertiesType {
    return (
        CalloutProperties.is(datum) ||
        CommentProperties.is(datum) ||
        NoteProperties.is(datum) ||
        TextProperties.is(datum)
    );
}
