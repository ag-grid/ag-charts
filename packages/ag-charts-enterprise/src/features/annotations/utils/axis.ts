import type { _ModuleSupport } from 'ag-charts-community';

export function calculateAxisLabelPadding(axisLayout: _ModuleSupport.AxisLayout) {
    return axisLayout.gridPadding + axisLayout.seriesAreaPadding + axisLayout.tickSize + axisLayout.label.padding;
}
