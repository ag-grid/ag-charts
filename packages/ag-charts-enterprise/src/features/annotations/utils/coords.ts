import { _ModuleSupport, _Scene } from 'ag-charts-community';

import type { AnnotationContext, Point } from '../annotationTypes';
import { convertPoint, invertCoords } from './values';

const { Vec2 } = _ModuleSupport;
const { toRadians } = _Scene;

export function snapPoint(
    offset: _ModuleSupport.Vec2,
    context: AnnotationContext,
    snapping: boolean = false,
    origin?: Point,
    angleStep: number = 1
) {
    if (!snapping) return invertCoords(offset, context);

    const center = origin ? convertPoint(origin, context) : Vec2.origin();
    return invertCoords(snapToAngle(offset, center, angleStep), context);
}

export function snapToAngle(
    { x, y }: _ModuleSupport.Vec2,
    center: _ModuleSupport.Vec2,
    step: number,
    direction: number = 1
) {
    const { x: cx, y: cy } = center;
    const r = Math.sqrt(Math.pow(x - cx, 2) + Math.pow(y - cy, 2));
    const theta = Math.atan2(y - cy, x - cx);

    const stepRadians = toRadians(step);
    const snapTheta = Math.round(theta / stepRadians) * stepRadians;

    return {
        x: cx + r * Math.cos(snapTheta),
        y: cy + r * Math.sin(snapTheta) * direction,
    };
}
