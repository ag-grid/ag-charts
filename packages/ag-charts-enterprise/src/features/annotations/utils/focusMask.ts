import { _ModuleSupport, _Util } from 'ag-charts-community';

const { partialAssign, getWindow, createElement } = _ModuleSupport;
const { copyAttributes } = _Util;

export function createFocusMask(target: HTMLElement): HTMLElement {
    const mask = createElement('div');
    const targetStyle = getWindow().getComputedStyle(target);

    mask.className = 'ag-charts-focus-mask';
    mask.tabIndex = -1;
    mask.style.position = 'absolute';
    mask.style.pointerEvents = 'none';
    partialAssign(['width', 'height'], mask.style, targetStyle);
    copyAttributes(['role', 'aria-labelledby'], mask, target);

    const disappear = () => {
        mask.removeEventListener('keydown', disappear);
        mask.removeEventListener('blur', disappear);
        mask.remove();
    };
    mask.addEventListener('blur', disappear);
    mask.addEventListener('keydown', () => {
        target.focus();
        disappear();
    });

    target.insertAdjacentElement('beforebegin', mask);
    return mask;
}
