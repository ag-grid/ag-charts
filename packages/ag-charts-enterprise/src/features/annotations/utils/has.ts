import { _ModuleSupport } from 'ag-charts-community';

import type {
    AnnotationProperties,
    ChannelPropertiesType,
    LinePropertiesType,
    MeasurerPropertiesType,
    TextualPropertiesType,
} from '../annotationsSuperTypes';
import { CalloutProperties } from '../callout/calloutProperties';
import { CommentProperties } from '../comment/commentProperties';
import { NoteProperties } from '../note/noteProperties';
import { ShapePointProperties } from '../properties/shapePointProperties';
import { isChannelType, isLineType, isMeasurerType, isTextType } from './types';

const { isObject } = _ModuleSupport;

export function hasFontSize(datum?: AnnotationProperties): datum is Exclude<TextualPropertiesType, NoteProperties> {
    return isTextType(datum) && !NoteProperties.is(datum);
}

export function hasLineStyle(
    datum?: AnnotationProperties
): datum is LinePropertiesType | ChannelPropertiesType | MeasurerPropertiesType {
    return isLineType(datum) || isChannelType(datum) || isMeasurerType(datum);
}

export function hasLineColor(datum?: AnnotationProperties) {
    return (
        isLineType(datum) ||
        isChannelType(datum) ||
        isMeasurerType(datum) ||
        CalloutProperties.is(datum) ||
        NoteProperties.is(datum)
    );
}

export function hasIconColor(datum?: AnnotationProperties) {
    return NoteProperties.is(datum);
}

export function hasFillColor(datum?: AnnotationProperties) {
    return (
        isChannelType(datum) ||
        isMeasurerType(datum) ||
        CalloutProperties.is(datum) ||
        CommentProperties.is(datum) ||
        ShapePointProperties.is(datum)
    );
}

export function hasTextColor(datum?: AnnotationProperties) {
    return isTextType(datum) && !NoteProperties.is(datum);
}

export function hasLineText(datum?: AnnotationProperties): datum is LinePropertiesType | ChannelPropertiesType {
    return (isLineType(datum) || isChannelType(datum) || isMeasurerType(datum)) && isObject(datum.text);
}
