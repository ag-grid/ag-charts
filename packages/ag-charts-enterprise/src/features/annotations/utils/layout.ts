import { _Scene } from 'ag-charts-community';

export type PositionedScene = Vec2Scene | Vec4Scene;

type Vec2Scene = _Scene.Node & { x: number; y: number };
type Vec4Scene = _Scene.Node & { x: number; y: number; x1: number; y1: number; x2: number; y2: number };

export function layoutScenesRow(
    scenes: Array<PositionedScene | Array<PositionedScene>>,
    startX: number = 0,
    gap: number = 0
) {
    let x = startX;

    for (const scene of scenes) {
        if (Array.isArray(scene)) {
            for (const scene_ of scene) {
                layoutSetX(scene_, x);
            }
            x += _Scene.Group.computeChildrenBBox(scene).width + gap;
        } else {
            layoutSetX(scene, x);
            x += scene.getBBox().width + gap;
        }
    }
}

export function layoutScenesColumn(
    scenes: Array<PositionedScene | Array<PositionedScene>>,
    startY: number = 0,
    gap: number = 0
) {
    let y = startY;

    for (const scene of scenes) {
        if (Array.isArray(scene)) {
            for (const scene_ of scene) {
                layoutSetY(scene_, y);
            }
            y += _Scene.Group.computeChildrenBBox(scene).height + gap;
        } else {
            layoutSetY(scene, y);
            y += scene.getBBox().height + gap;
        }
    }
}

export function layoutSetX(scene: PositionedScene, x: number) {
    if ('x1' in scene) {
        scene.x2 = x + (scene.x2 - scene.x1);
        scene.x1 = x;
    } else {
        scene.x = x;
    }
}

export function layoutSetY(scene: PositionedScene, y: number) {
    if ('y1' in scene) {
        scene.y2 = y + (scene.y2 - scene.y1);
        scene.y1 = y;
    } else {
        scene.y = y;
    }
}

export function layoutAddX(scene: PositionedScene, x: number) {
    if ('x1' in scene) {
        scene.x1 += x;
        scene.x2 += x;
    } else {
        scene.x += x;
    }
}

export function layoutAddY(scene: PositionedScene, y: number) {
    if ('y1' in scene) {
        scene.y1 += y;
        scene.y2 += y;
    } else {
        scene.y += y;
    }
}
