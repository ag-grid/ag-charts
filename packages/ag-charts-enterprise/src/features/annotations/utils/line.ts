import type { AgAnnotationLineStyleType, PixelSize, _Scene } from 'ag-charts-community';

export function getLineStyle(lineDash?: PixelSize[], lineStyle?: AgAnnotationLineStyleType) {
    return lineDash ? 'dashed' : lineStyle ?? 'solid';
}

export function getComputedLineDash(strokeWidth: number, styleType: AgAnnotationLineStyleType): PixelSize[] {
    switch (styleType) {
        case 'solid':
            return [];
        case 'dashed':
            return [strokeWidth * 4, strokeWidth * 2];
        case 'dotted':
            return [0, strokeWidth * 2];
    }
}

export function getLineDash(
    lineDash?: PixelSize[],
    computedLineDash?: PixelSize[],
    lineStyle?: AgAnnotationLineStyleType,
    strokeWidth?: number
): PixelSize[] | undefined {
    const styleType = getLineStyle(lineDash, lineStyle);
    return computedLineDash ?? lineDash ?? getComputedLineDash(strokeWidth ?? 1, styleType);
}

export function getLineCap(
    lineCap?: _Scene.ShapeLineCap,
    lineDash?: PixelSize[],
    lineStyle?: AgAnnotationLineStyleType
): _Scene.ShapeLineCap | undefined {
    const styleType = getLineStyle(lineDash, lineStyle);
    return lineCap ?? styleType === 'dotted' ? 'round' : undefined;
}
