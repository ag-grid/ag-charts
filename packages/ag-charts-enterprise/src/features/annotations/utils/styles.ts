import type { AnnotationLineStyle, AnnotationOptionsColorPickerType } from '../annotationTypes';
import type { AnnotationProperties, ChannelPropertiesType, LinePropertiesType } from '../annotationsSuperTypes';
import { hasIconColor, hasLineText } from './has';
import { getComputedLineDash, getLineStyle } from './line';

export function setFontSize(datum: AnnotationProperties, fontSize: number) {
    if ('fontSize' in datum) datum.fontSize = fontSize;
    if (hasLineText(datum)) datum.text.fontSize = fontSize;
}

export function setLineStyle(datum: LinePropertiesType | ChannelPropertiesType, style?: AnnotationLineStyle) {
    const strokeWidth = style?.strokeWidth ?? datum.strokeWidth ?? 1;
    const lineType = style?.type ?? datum.lineStyle;
    const lineStyle = lineType ?? getLineStyle(datum.lineDash, lineType);
    const computedLineDash = getComputedLineDash(strokeWidth, lineStyle);

    datum.strokeWidth = strokeWidth;
    datum.computedLineDash = computedLineDash;
    datum.lineStyle = lineStyle;
    datum.lineCap = lineStyle === 'dotted' ? 'round' : undefined;
}

export function setColor(
    datum: AnnotationProperties,
    colorPickerType: AnnotationOptionsColorPickerType,
    colorOpacity: string,
    color: string,
    opacity: number
) {
    switch (colorPickerType) {
        case `fill-color`: {
            if ('fill' in datum) datum.fill = color;
            if ('fillOpacity' in datum) datum.fillOpacity = opacity;
            if ('background' in datum) {
                datum.background.fill = color;
                datum.background.fillOpacity = opacity;
            }
            break;
        }

        case `line-color`: {
            if ('axisLabel' in datum) {
                datum.axisLabel.fill = color;
                datum.axisLabel.fillOpacity = opacity;
                datum.axisLabel.stroke = color;
                datum.axisLabel.strokeOpacity = opacity;
            }

            if ('fill' in datum && 'fillOpacity' in datum && hasIconColor(datum)) {
                datum.fill = color;
                datum.fillOpacity = opacity;
            } else {
                if ('stroke' in datum) datum.stroke = color;
                if ('strokeOpacity' in datum) datum.strokeOpacity = opacity;
            }

            break;
        }

        case `text-color`: {
            if ('color' in datum) datum.color = colorOpacity;
            if (hasLineText(datum)) datum.text.color = color;
            break;
        }
    }
}
