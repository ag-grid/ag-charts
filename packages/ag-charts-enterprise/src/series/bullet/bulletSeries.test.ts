import { afterEach, describe, expect, it, xit } from '@jest/globals';

import { type AgChartInstance, AgCharts } from 'ag-charts-community';
import {
    IMAGE_SNAPSHOT_DEFAULTS,
    deproxy,
    expectWarning,
    extractImageData,
    hoverAction,
    setupMockCanvas,
    setupMockConsole,
    waitForChartStability,
} from 'ag-charts-community-test';

import { prepareEnterpriseTestOptions } from '../../test/utils';

type TChart = Parameters<typeof waitForChartStability>[0];
type TCtx = ReturnType<typeof setupMockCanvas>;

const compare = async (chart: TChart | AgChartInstance | undefined, ctx: TCtx) => {
    expect(chart).toBeDefined();
    if (chart === undefined) return;

    await waitForChartStability(chart as TChart);
    const imageData = extractImageData(ctx);
    expect(imageData).toMatchImageSnapshot({ ...IMAGE_SNAPSHOT_DEFAULTS, failureThreshold: 0 });
};

describe('BulletSeries', () => {
    setupMockConsole();

    let chart: AgChartInstance | undefined;
    const ctx = setupMockCanvas();

    afterEach(() => {
        chart?.destroy();
        chart = undefined;
    });

    const opts = prepareEnterpriseTestOptions({});

    describe('rendering', () => {
        const getTooltipHtml = (): string => {
            const series = (chart as any)['series'][0];
            const datum = (chart as any)['series'][0].contextNodeData?.nodeData[0];
            return series.getTooltipHtml(datum).html;
        };

        const hoverOnBullet = async () => {
            const series = chart['series'][0];
            const item = series['contextNodeData'].nodeData[0];
            const { x, y } = series.rootGroup.inverseTransformPoint(item.midPoint.x, item.midPoint.y);
            await hoverAction(x, y)(chart);
        };

        it('should render simple bullet', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [
                    {
                        type: 'bullet',
                        data: [{ income: 11, objective: 7 }],
                        valueKey: 'income',
                    },
                ],
            });
            await compare(chart, ctx);
        });

        it('should render a vertical bullet by default', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [
                    {
                        type: 'bullet',
                        data: [{ income: 11, objective: 7 }],
                        valueKey: 'income',
                        valueName: 'Actual income',
                        targetKey: 'objective',
                        targetName: 'Target income',
                        colorRanges: [{ color: 'red', stop: 5 }, { color: 'yellow', stop: 8 }, { color: 'green' }],
                    },
                ],
            });
            await compare(chart, ctx);
        });

        it('should render a horizontal bullet as expected', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [
                    {
                        type: 'bullet',
                        data: [{ income: 11, objective: 7 }],
                        valueKey: 'income',
                        valueName: 'Actual income',
                        targetKey: 'objective',
                        targetName: 'Target income',
                        direction: 'horizontal',
                        colorRanges: [{ color: 'red', stop: 5 }, { color: 'yellow', stop: 8 }, { color: 'green' }],
                    },
                ],
            });
            await compare(chart, ctx);
        });

        it('should clip everything to scale.max', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [
                    {
                        type: 'bullet',
                        data: [{ income: 11, objective: 7 }],
                        scale: { max: 5 },
                        valueKey: 'income',
                        targetKey: 'objective',
                        colorRanges: [{ color: 'yellow', stop: 45 }],
                    },
                ],
            });
            await compare(chart, ctx);
        });

        it('should extend final color to scale.max', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [
                    {
                        type: 'bullet',
                        data: [{ income: 11, objective: 7 }],
                        scale: { max: 20 },
                        valueKey: 'income',
                        targetKey: 'objective',
                        colorRanges: [{ color: 'red', stop: 5 }, { color: 'yellow', stop: 8 }, { color: 'green' }],
                    },
                ],
            });
            await compare(chart, ctx);
        });

        it('should use explicit axis max', async () => {
            chart = AgCharts.create({
                ...opts,
                axes: [{ type: 'number', max: 50 }, { type: 'category' }],
                series: [
                    {
                        type: 'bullet',
                        data: [{ income: 11 }],
                        scale: { max: 20 },
                        valueKey: 'income',
                    },
                ],
            });
            await compare(chart, ctx);
        });

        it('should process first datum only', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [
                    {
                        type: 'bullet',
                        data: [{ income: 1 }, { income: 2 }, { income: 3 }],
                        scale: { max: 1.5 },
                        valueKey: 'income',
                    },
                ],
            });
            await compare(chart, ctx);
        });

        it('should not render crosshair', async () => {
            chart = deproxy(
                AgCharts.create({
                    ...opts,
                    series: [
                        {
                            type: 'bullet',
                            data: [{ income: 1 }],
                            scale: { max: 2 },
                            valueKey: 'income',
                        },
                    ],
                })
            );
            await waitForChartStability(chart);
            await hoverOnBullet();
            await compare(chart, ctx);
        });

        test('tooltip valueKey only', async () => {
            chart = deproxy(
                AgCharts.create({
                    ...opts,
                    series: [
                        {
                            type: 'bullet',
                            valueKey: 'income',
                            data: [{ income: 11, objective: 7 }],
                        },
                    ],
                })
            );
            await waitForChartStability(chart);

            const tooltipHtml = getTooltipHtml();
            expect(tooltipHtml).toEqual('<div class="ag-chart-tooltip-content"><b>income</b>: 11.0</div>');
        });

        test('tooltip no names', async () => {
            chart = deproxy(
                AgCharts.create({
                    ...opts,
                    series: [
                        {
                            type: 'bullet',
                            valueKey: 'income',
                            targetKey: 'objective',
                            data: [{ income: 11, objective: 7 }],
                        },
                    ],
                })
            );
            await waitForChartStability(chart);

            const tooltipHtml = getTooltipHtml();
            expect(tooltipHtml).toEqual(
                '<div class="ag-chart-tooltip-content"><b>income</b>: 11.0<br/><b>objective</b>: 7.0</div>'
            );
        });

        test('tooltip with names', async () => {
            chart = deproxy(
                AgCharts.create({
                    ...opts,
                    series: [
                        {
                            type: 'bullet',
                            valueKey: 'income',
                            valueName: 'Actual income',
                            targetKey: 'objective',
                            targetName: 'Target income',
                            data: [{ income: 11, objective: 7 }],
                        },
                    ],
                })
            );
            await waitForChartStability(chart);

            const tooltipHtml = getTooltipHtml();
            expect(tooltipHtml).toEqual(
                '<div class="ag-chart-tooltip-content"><b>Actual income</b>: 11.0<br/><b>Target income</b>: 7.0</div>'
            );
        });
    });

    describe('validation', () => {
        // Disabled as a consequence of AG-10326
        xit('should ignore empty colorRange arrays', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [
                    {
                        type: 'bullet',
                        data: [{ income: 11 }],
                        scale: { max: 20 },
                        valueKey: 'income',
                        colorRanges: [],
                    },
                ],
            });
            await waitForChartStability(chart);

            expectWarning(
                'AG Charts - Property [colorRanges] of [BulletSeries] cannot be set to [[]]; expecting a non-empty array, ignoring.'
            );
            await compare(chart, ctx);
        });

        it('should ignore trailing bullet series', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [
                    {
                        type: 'bullet',
                        data: [{ income: 78, objective: 90 }],
                        scale: { max: 100 },
                        valueKey: 'income',
                        targetKey: 'objective',
                    },
                    {
                        type: 'bullet',
                        data: [{ income: 234 }],
                        scale: { max: 300 },
                        valueKey: 'income',
                        colorRanges: [{ color: 'red' }],
                    },
                    {
                        type: 'bullet',
                        data: [{ cost: 30, costTarget: 60 }],
                        scale: { max: 120 },
                        valueKey: 'cost',
                        targetKey: 'costTarget',
                        fill: 'blue',
                        target: { stroke: 'green' },
                    },
                ],
            });
            await waitForChartStability(chart);

            expectWarning(
                `AG Charts - series[0] of type 'bullet' is incompatible with other series types. Only processing series[0]`
            );
            await compare(chart, ctx);
        });

        it('should ignore other trailing series', async () => {
            chart = AgCharts.create({
                ...opts,
                data: [
                    { a: 23, b: 17, yr: '1990' },
                    { a: 53, b: 59, yr: '2000' },
                    { a: 47, b: 65, yr: '2010' },
                ],
                series: [
                    {
                        type: 'bullet',
                        data: [{ cost: 30, costTarget: 60 }],
                        scale: { max: 120 },
                        valueKey: 'cost',
                        targetKey: 'costTarget',
                        fill: 'blue',
                        target: { stroke: 'green' },
                    },
                    { type: 'line', xKey: 'yr', yKey: 'a' },
                    { type: 'bar', xKey: 'yr', yKey: 'b' },
                ],
            });
            await waitForChartStability(chart);

            expectWarning(
                `AG Charts - series[0] of type 'bullet' is incompatible with other series types. Only processing series[0]`
            );
            await compare(chart, ctx);
        });

        it('should filter bullet series from line chart', async () => {
            chart = AgCharts.create({
                ...opts,
                data: [
                    { a: 23, b: 17, yr: '1990' },
                    { a: 53, b: 59, yr: '2000' },
                    { a: 47, b: 65, yr: '2010' },
                ],
                series: [
                    { type: 'line', xKey: 'yr', yKey: 'a' },
                    { type: 'line', xKey: 'yr', yKey: 'b' },

                    {
                        type: 'bullet',
                        data: [{ income: 78 }],
                        scale: { max: 100 },
                        valueKey: 'income',
                    },
                    {
                        type: 'bullet',
                        data: [{ cost: 30, costTarget: 60 }],
                        scale: { max: 120 },
                        valueKey: 'cost',
                        targetKey: 'costTarget',
                        fill: 'blue',
                        target: { stroke: 'green' },
                    },
                ],
            });
            await waitForChartStability(chart);

            expectWarning('AG Charts - Unable to mix these series types with the lead series type: bullet');
            await compare(chart, ctx);
        });

        test('warning negative value', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [{ type: 'bullet', data: [{ v: -1 }], valueKey: 'v' }],
            });
            await waitForChartStability(chart);

            expectWarning('AG Charts - negative values are not supported, clipping to 0.');
            await compare(chart, ctx);
        });

        test('warning negative target', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [{ type: 'bullet', data: [{ v: 11, t: -1 }], valueKey: 'v', targetKey: 't' }],
            });
            await waitForChartStability(chart);

            expectWarning('AG Charts - negative targets are not supported, ignoring.');
            await compare(chart, ctx);
        });

        test('warning negative max, ignore max', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [{ type: 'bullet', data: [{ v: 11 }], valueKey: 'v', scale: { max: -1 } }],
            });
            await waitForChartStability(chart);

            expectWarning(
                'AG Charts - Property [max] of [BulletScale] cannot be set to [-1]; expecting a number greater than or equal to 0, ignoring.'
            );
            await compare(chart, ctx);
        });

        test('warning negative colorRange stop', async () => {
            chart = AgCharts.create({
                ...opts,
                series: [
                    { type: 'bullet', data: [{ v: 11 }], valueKey: 'v', colorRanges: [{ color: 'blue', stop: -1 }] },
                ],
            });
            await waitForChartStability(chart);

            expectWarning(
                'AG Charts - Property [stop] of [BulletColorRange] cannot be set to [-1]; expecting a number greater than or equal to 0, ignoring.'
            );
            await compare(chart, ctx);
        });
    });
});
