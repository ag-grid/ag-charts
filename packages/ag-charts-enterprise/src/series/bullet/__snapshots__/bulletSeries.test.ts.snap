// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulletSeries rendering tooltip no names 1`] = `"<div class="ag-chart-tooltip-content"><b>income</b>: 11.0<br><b>objective</b>: 7.0</div>"`;

exports[`BulletSeries rendering tooltip with names 1`] = `"<div class="ag-chart-tooltip-content"><b>Actual income</b>: 11.0<br><b>Target income</b>: 7.0</div>"`;
